NN Archive - Parsers Support

kkeroo opened the pull request:
This PR adds support for parsers for all heads. Now, each head has `parser_name` and `metadata` that stores key-value pairs for required parameters. Parameters from the old head are moved to `metadata` object (e.g. `n_keypoints`, `classes`,...).

Each head in NN archive has its own `Metadata` model. Starting with the basic head having `HeadMetadata`, then `HeadObjectDetection` having `HeadObjectDetectionMetadata`. All `Metadata` models also allow you to have custom fields, but they still validate the required ones. The only problem here is that code editor still underlines the custom parameter as unsupported, you can avoid this by adding `#type: ignore`. Does anyone know a better way?
```
metadata = HeadMetadata(
    classes=['person', 'car'],
    n_classes=2,
    custom_parameter='custom', # type: ignore
)
```

YOLO heads still check the right combination of parameters and allow multiple tasks.

Additionally, one test is added for testing the custom parameter setting in `Metadata` object.

One example of `config.json` with custom parameter:
```
{
    "config_version": "1.0",
    "model": {
        "metadata": {
            "name": "XFeat",
            "path": "matching.onnx",
            "precision": "float32"
        },
        "inputs": [
            {
                "name": "images",
                "dtype": "float32",
                "input_type": "image",
                "shape": [
                    1,
                    3,
                    352,
                    640
                ],
                "layout": "NCHW",
                "preprocessing": {
                    "mean": [
                        0.0,
                        0.0,
                        0.0
                    ],
                    "scale": [
                        255.0,
                        255.0,
                        255.0
                    ],
                    "reverse_channels": false,
                    "interleaved_to_planar": false
                }
            }
        ],
        "outputs": [
            {
                "name": "output",
                "dtype": "float32"
            }
        ],
        "heads": [
            {
                "family": "Segmentation",
                "parser_name": "PP-lites",
                "metadata": {
                    "classes": [
                        "person",
                        "car"
                    ],
                    "n_classes": 2,
                    "is_softmax": false,
                    "custom_field": "custom_field"
                },
                "outputs": {
                    "predictions": "output"
                }
            }
        ]
    }
}
```

kkeroo commented:
Dataset test is failing, this has nothing to do with the changes.

kozlov721 commented on luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:
```diff
@@ -144,6 +163,15 @@ def validate_label_type(
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadSegmentationMetadata):
```
Can this ever happen with the `pydantic` validation already in place?

kkeroo replied:
Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

kozlov721 commented on luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:
```diff
@@ -20,51 +27,42 @@ class Head(BaseModel, ABC):
+
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadMetadata):
```
Can this ever happen with the `pydantic` validation already in place?

kkeroo replied:
Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

kozlov721 commented on luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:
```diff
@@ -93,6 +91,15 @@ def validate_label_type(
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadClassificationMetadata):
```
Same as above

kkeroo replied:
Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

kozlov721 commented on luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:
```diff
@@ -117,6 +127,15 @@ def validate_label_type(
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadObjectDetectionMetadata):
```
Same as above

kkeroo replied:
Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

kozlov721 commented on luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:
```diff
@@ -195,12 +199,23 @@ def validate_label_type(
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadYOLOMetadata):
```
Same as above

kkeroo replied:
Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

kozlov721 reviewed (approved):
LGTM

kkeroo commented:
### Update

We removed `family` from all heads and replaced it with `parser`. Also, `postprocessor_path` field is in every head and removed from Yolo's head metadata.

Every head now looks like:
```
Head:
  parser
  postprocessor_path
  metadata
  outputs
```

tersekmatija reviewed (approved):
Looks good, let's make the following optional on the Inputs so we can deprecate it inside DAI and make more explicit later:
```
                    "reverse_channels": true,
                    "interleaved_to_planar": false
```

kkeroo commented:
### Update

1. `postprocesor_path` back in `Metadata`
2. Optional `reverse_input_channels` and `interleaved_to_planar`
3. `outputs` on the head level only lists of strings
4. `outputs` in `Metadata` only for `YOLO` and `SSD`
5. validation moved to `Metadata`
6. only a single `Head`
